README
NOTE: Although icecream will compile on some non-Linux systems,
it depends critically on Linux' /proc/ structures and will fail
at runtime until somebody figures out how to emulate the "ticks"
form of process accounting in there.

How to install icecream
=======================

cd icecream
./autogen.sh
./configure --prefix=/opt/icecream
make
make install

WARNING: Make sure you specify a prefix, otherwise icecream
might override your gcc installation! You will need to use this
prefix instead of /usr when referring to icecream (for example
when extending the $PATH variable).

The documentation is maintained in README.md for the time being.

Host to build/install with cmake
================================

WARNING, the cmake build is a work in progress, it probably doesn't even 
complete. Even if you somehow get a successful build that does not mean that
the build actually will work. There are many known issues.  Bug reports will
not be accepted.

cd icecream
mkdir build
cd build
cmake -DCMAKE_INSTALL_PREFIX=/opt/icecream -DBUILD_EXPIRIMENTAL=true ..
make -j 4

How to contribute
=================

Fork on github.com and send a pull request:
https://github.com/icecc/icecream

There is a testsuite in the tests/ directory, see the README file there.

Mailing list:
icecream-users@googlegroups.com  (icecream-users+subscribe@googlegroups.com)


NEWS
1.2 (in progress)
<<<<<<< HEAD
=======
C++11 is required
 - more logging, entries now have date as well as timestamp.
>>>>>>> Used pkgconfig to find libcap-ng

1.1
 - revert "Add load control for preprocessing"
 - better handle clang arugments with spaces
 - remove "crashme" command from scheduler
 - better logging around exception 30

1.1rc3
 - Fix broken pipe race condition
 - Better error handling
 - Crash fixes
 - Documentation update
 - Include objcopy in environment if it exists
 - Add CI builds on travis-ci: OSX, ubuntu trusty
 - Fixed several memory errors
 - Extract enviornments with compile priority, not daemon priority
 - Handle spaces in the parameters following -MT
 - Handle -target -arch -c-isystem and -cxx-isystem (clang)
 - Handle NAT situations better
 - Add load control for preprocessing
 - Handle scheduler unable to reach remote machine
 - Make scheduler election algorithm handle multiple netnames

1.1rc2
 - -gsplit-dwarf support for debug fission (https://gcc.gnu.org/wiki/DebugFission)
 - bug fixes since 1.1rc1

1.1rc1
 - require capng
 - use system lzo, drop bundled minilzo
 - allow ICECC_VERSION file to start with a .
 - work with different executable names for docbook-to-man on different distributions
 - support color diagnostics if possible
 - fix several crashes
 - clean up and improve documentation
 - move icecc-create-env to bindir.
 - Try to use only the best available scheduler
 - make daemon port configurable
 - USE_SCHEDULER now accepts host:port to change ports
 - force local compilation when required in more cases
 - detect some clang out of memory errors.
 - better support for FreeBSD
 - bump protocol version to 34
 - daemon requires chroot
 - some code refactoring for better maintainability
 - find if feature is supported by checking protocol version
 - better logging on error conditions
 - Create many tests cases
 - use path name instead of localhost for unix socket
 - clean up build system
 - use getnameinfo() instead of inet_ntoa()
 - fix some valgrind found issues
 - drop supplementary groups before setgid()
 - make signal handlers more robust
 - work better if user "icecc" does not exist
 - find compilers outside of PATH when building locally

1.0.1
- use su option for logrotate
- require logrotate for suse package

1.0.1rc1
- Remove filenames from md5sum output
- Fix off-by-one error
- add missing cap-ng.m4
- uninstall also SUSE-specific files when uninstalling
- install COPYING, README.md and NEWS as useful user documentation
- use m4/ and ship cap-ng.m4 there
- if -lcap-ng is needed for icecc.a, add it to .pc file (for icemon)
- libicecream-devel may require libcap-ng-devel
- fix icecc-create-env with clang 3.3+
- fix /var/run/icecc handling
- make sure to not hide e.g. SEGV from a tool with exit code 0
- detect bad_alloc as out-of-memory
- add "[Cc]annot allocate memory" to the list of OOM error messages
- Detect gcc "out of memory allocating" as EXIT_OUT_OF_MEMORY. Closes #49
- Avoid symlinks in the absolute paths
- swap arguments for kill() call, since pid is the first argument

1.0.0
- log error message when cleaning up cache fails during startup
- if getuid() != 0, then our u/gid will be getu/gid() the whole time
- chmod/chown envs dir when preparing this
- be more careful when cleaning up envs directory
- cleanup envs dir still with root privileges
- mkdir -p
- Revert "cleanup envs dir still with root privileges"
- cleanup envs dir in %post in specfile
- fix scheduler binary name in suse specfile
- with cap-ng geteuid() is not a sign of being able to do chroot
- avoid debug message without endl
- avoid unused parameter warnings
- install clang/clang++ wrappers symlinks conditionally again
- adjust specfile for optional clang symlinks
- make clang wrappers package also require clang
- fix clang wrappers build
- refer to icecream-users@googlegroups.com ML as the contact
- mention the Linux3_ hack leading to nodes sometimes not being used
- Linux3_ nodes may not compile on other nodes either, actually
- prevent icerun-wrapped commands failing if path contains 'clang'
- remove unnecessary references to KDE to make icecream look KDE-specific
- adjust references to icecream path
- sync the ccache section between README.md and the manpage
- AC_CONFIG_HEADERS instead of obsolete and removed AM_CONFIG_HEADER
- add the README.md to the tar
- mild relicensing, the code taken from ksysguardd is really a minor part by now

0.9.98.2 (1.0rc4)
- suse: Backports from OBS and create symlinks in /opt/icecream/bin on openSUSE <= 12.2.
- fix libexec dir name in opensuse specfile
- icecream doesn't use m4 directory
- explicit configure option for whether to use clang's include rewriting
- use the clang rewrite options in opensuse specfile
- fix builddir != srcdir
- icecream user on suse is 'icecream', not 'icecc'
- do not remove env. basedir itself
- do not chown() with root uid
- do not complain needlessly about missing icecc user
- Fix icecc-create-env for relative paths
- Use docbook2man to generate manpages rather than KDE calls.

0.9.98.1 (1.0rc3)
 - install clang++ clang symlinks unconditionally.
 - move openSUSE's rpm install scripts to suse/Makefile.am.
 - links installation should be pkglibexec instead of libexec
 - rename scheduler to icecc-scheduler to avoid name clash

services/logging.cpp
/* -*- mode: C++; indent-tabs-mode: nil; c-basic-offset: 4; fill-column: 99; -*- */
/* vim: set ts=4 sw=4 et tw=99:  */
/*
    This file is part of Icecream.

    Copyright (c) 2004 Stephan Kulow <coolo@suse.de>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/

#include <config.h>
#include <iostream>
#include "logging.h"
#include <fstream>
#include <signal.h>
#include <limits.h>
#include <fcntl.h>
#ifdef __linux__
#include <dlfcn.h>
#endif

using namespace std;

int debug_level = Error;
ostream *logfile_trace = 0;
ostream *logfile_info = 0;
ostream *logfile_warning = 0;
ostream *logfile_error = 0;
string logfile_prefix;
volatile sig_atomic_t reset_debug_needed = 0;

static ofstream logfile_null("/dev/null");
static ofstream logfile_file;
static string logfile_filename;

static void reset_debug_signal_handler(int);

// Implementation of an iostream helper that allows redirecting output to a given file descriptor.
// This seems to be the only portable way to do it.
namespace
{
class ofdbuf : public streambuf
{
public:
    explicit ofdbuf( int fd ) : fd( fd ) {}
    virtual int_type overflow( int_type c );
    virtual streamsize xsputn( const char* c, streamsize n );
private:
    int fd;
};

ofdbuf::int_type ofdbuf::overflow( int_type c )
{
    if( c != EOF ) {
        char cc = c;
        if( write( fd, &cc, 1 ) != 1 )
            return EOF;
    }
    return c;
}

streamsize ofdbuf::xsputn( const char* c, streamsize n )
{
    return write( fd, c, n );
}

ostream* ccache_stream( int fd )
{
    int status = fcntl( fd, F_GETFL );
    if( status < 0 || ( status & ( O_WRONLY | O_RDWR )) == 0 ) {
        // As logging is not set up yet, this will log to stderr.
        log_warning() << "UNCACHED_ERR_FD provides an invalid file descriptor, using stderr" << endl;
        return &cerr; // fd is not valid fd for writting
    }
    static ofdbuf buf( fd );
    static ostream stream( &buf );
    return &stream;
}
} // namespace

void setup_debug(int level, const string &filename, const string &prefix)
{
    debug_level = level;
    logfile_prefix = prefix;
    logfile_filename = filename;

    if (logfile_file.is_open()) {
        logfile_file.close();
    }

    ostream *output = 0;

    if (filename.length()) {
        logfile_file.clear();
        logfile_file.open(filename.c_str(), fstream::out | fstream::app);
#ifdef __linux__

        string fname = filename;
        if (fname[0] != '/') {
            char buf[PATH_MAX];

            if (getcwd(buf, sizeof(buf))) {
                fname.insert(0, "/");
                fname.insert(0, buf);
            }
        }

        setenv("SEGFAULT_OUTPUT_NAME", fname.c_str(), false);
#endif
        output = &logfile_file;
    } else if( const char* ccache_err_fd = getenv( "UNCACHED_ERR_FD" )) {
        output = ccache_stream( atoi( ccache_err_fd ));
    } else {
        output = &cerr;
    }

#ifdef __linux__
    (void) dlopen("libSegFault.so", RTLD_NOW | RTLD_LOCAL);
#endif

    if (debug_level >= Debug) {
        logfile_trace = output;
    } else {
        logfile_trace = &logfile_null;
    }

    if (debug_level >= Info) {
        logfile_info = output;
    } else {
        logfile_info = &logfile_null;
    }

    if (debug_level >= Warning) {
        logfile_warning = output;
    } else {
        logfile_warning = &logfile_null;
    }

    if (debug_level >= Error) {
        logfile_error = output;
    } else {
        logfile_error = &logfile_null;
    }

    signal(SIGHUP, reset_debug_signal_handler);
}

void reset_debug()
{
    setup_debug(debug_level, logfile_filename);
}

void reset_debug_signal_handler(int)
{
    reset_debug_needed = 1;
}

void reset_debug_if_needed()
{
    if( reset_debug_needed ) {
        reset_debug_needed = 0;
        reset_debug();
        if( const char* env = getenv( "ICECC_TEST_FLUSH_LOG_MARK" )) {
            ifstream markfile( env );
            string mark;
            getline( markfile, mark );
            if( !mark.empty()) {
                assert( logfile_trace != NULL );
                *logfile_trace << "flush log mark: " << mark << endl;
            }
        }
        if( const char* env = getenv( "ICECC_TEST_LOG_HEADER" )) {
            ifstream markfile( env );
            string header1, header2, header3;
            getline( markfile, header1 );
            getline( markfile, header2 );
            getline( markfile, header3 );
            if( !header1.empty()) {
                assert( logfile_trace != NULL );
                *logfile_trace << header1 << endl;
                *logfile_trace << header2 << endl;
                *logfile_trace << header3 << endl;
            }
        }
    }
}

void close_debug()
{
    if (logfile_null.is_open()) {
        logfile_null.close();
    }

    if (logfile_file.is_open()) {
        logfile_file.close();
    }

    logfile_trace = logfile_info = logfile_warning = logfile_error = 0;
}

/* Flushes all ostreams used for debug messages.  Youd need to call
   this before forking.  */
void flush_debug()
{
    if (logfile_null.is_open()) {
        logfile_null.flush();
    }

    if (logfile_file.is_open()) {
        logfile_file.flush();
    }
}

unsigned log_block::nesting;


daemon/environment.cpp
/* -*- mode: C++; indent-tabs-mode: nil; c-basic-offset: 4; fill-column: 99; -*- */
/* vim: set ts=4 sw=4 et tw=99:  */
/*
    This file is part of Icecream.

    Copyright (c) 2004 Stephan Kulow <coolo@suse.de>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/

#include "config.h"
#include "environment.h"
#include <logging.h>
#include <errno.h>
#include <dirent.h>
#include <unistd.h>
#include <fcntl.h>
#include <grp.h>
#include <stdio.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <signal.h>

#include "comm.h"
#include "exitcode.h"
#include "util.h"

using namespace std;

size_t sumup_dir(const string &dir)
{
    size_t res = 0;
    DIR *envdir = opendir(dir.c_str());

    if (!envdir) {
        return res;
    }

    struct stat st;

    string tdir = dir + "/";

    for (struct dirent *ent = readdir(envdir); ent; ent = readdir(envdir)) {
        if (!strcmp(ent->d_name, ".") || !strcmp(ent->d_name, "..")) {
            continue;
        }

        if (lstat((tdir + ent->d_name).c_str(), &st)) {
            perror("stat");
            continue;
        }

        if (S_ISDIR(st.st_mode)) {
            res += sumup_dir(tdir + ent->d_name);
        } else if (S_ISREG(st.st_mode)) {
            res += st.st_size;
        }

        // else ignore
    }

    closedir(envdir);
    return res;
}

static void list_target_dirs(const string &current_target, const string &targetdir, Environments &envs)
{
    DIR *envdir = opendir(targetdir.c_str());

    if (!envdir) {
        return;
    }

    for (struct dirent *ent = readdir(envdir); ent; ent = readdir(envdir)) {
        string dirname = ent->d_name;

        if (!access(string(targetdir + "/" + dirname + "/usr/bin/as").c_str(), X_OK)) {
            envs.push_back(make_pair(current_target, dirname));
        }
    }

    closedir(envdir);
}

/* Returns true if the child exited with success */
static bool exec_and_wait(const char *const argv[])
{
    pid_t pid = fork();

    if (pid == -1) {
        log_perror("failed to fork");
        return false;
    }

    if (pid) {
        // parent
        int status;

        while (waitpid(pid, &status, 0) < 0 && errno == EINTR) {}

        return shell_exit_status(status) == 0;
    }

    // child
    execv(argv[0], const_cast<char * const *>(argv));
    log_perror("execv failed");
    _exit(-1);
}

// Removes everything in the directory recursively, but not the directory itself.
static bool cleanup_directory(const string &directory)
{
    DIR *dir = opendir(directory.c_str());

    if (dir == NULL) {
        return false;
    }

    while (dirent *f = readdir(dir)) {
        if (strcmp(f->d_name, ".") == 0 || strcmp(f->d_name, "..") == 0) {
            continue;
        }

        string fullpath = directory + '/' + f->d_name;
        struct stat st;

        if (lstat(fullpath.c_str(), &st)) {
            perror("stat");
            return false;
        }

        if (S_ISDIR(st.st_mode)) {
            if (!cleanup_directory(fullpath) || rmdir(fullpath.c_str()) != 0) {
                return false;
            }
        } else {
            if (unlink(fullpath.c_str()) != 0) {
                return false;
            }
        }
    }

    closedir(dir);
    return true;
}

bool cleanup_cache(const string &basedir, uid_t user_uid, gid_t user_gid)
{
    flush_debug();

    if (access(basedir.c_str(), R_OK) == 0 && !cleanup_directory(basedir)) {
        log_error() << "failed to clean up envs dir" << endl;
        return false;
    }

    if (mkdir(basedir.c_str(), 0755) && errno != EEXIST) {
        if (errno == EPERM) {
            log_error() << "permission denied on mkdir " << basedir << endl;
        } else {
            log_perror("mkdir in cleanup_cache() failed") << "\t" << basedir << endl;
        }

        return false;
    }

    if (chown(basedir.c_str(), user_uid, user_gid) || chmod(basedir.c_str(), 0775)) {
        log_perror("chown/chmod in cleanup_cache() failed") << "\t" << basedir << endl;;
        return false;
    }

    return true;
}

Environments available_environmnents(const string &basedir)
{
    Environments envs;

    DIR *envdir = opendir(basedir.c_str());

    if (!envdir) {
        log_info() << "can't open envs dir " << strerror(errno) << endl;
    } else {
        for (struct dirent *target_ent = readdir(envdir); target_ent; target_ent = readdir(envdir)) {
            string dirname = target_ent->d_name;

            if (dirname.at(0) == '.') {
                continue;
            }

            if (dirname.substr(0, 7) == "target=") {
                string current_target = dirname.substr(7, dirname.length() - 7);
                list_target_dirs(current_target, basedir + "/" + dirname, envs);
            }
        }

        closedir(envdir);
    }

    return envs;
}

void save_compiler_timestamps(time_t &gcc_bin_timestamp, time_t &gpp_bin_timestamp, time_t &clang_bin_timestamp)
{
    struct stat st;

    if (stat("/usr/bin/gcc", &st) == 0) {
        gcc_bin_timestamp = st.st_mtime;
    } else {
        gcc_bin_timestamp = 0;
    }

    if (stat("/usr/bin/g++", &st) == 0) {
        gpp_bin_timestamp = st.st_mtime;
    } else {
        gpp_bin_timestamp = 0;
    }

    if (stat("/usr/bin/clang", &st) == 0) {
        clang_bin_timestamp = st.st_mtime;
    } else {
        clang_bin_timestamp = 0;
    }
}

bool compilers_uptodate(time_t gcc_bin_timestamp, time_t gpp_bin_timestamp, time_t clang_bin_timestamp)
{
    struct stat st;

    if (stat("/usr/bin/gcc", &st) == 0) {
        if (st.st_mtime != gcc_bin_timestamp) {
            return false;
        }
    } else {
        if (gcc_bin_timestamp != 0) {
            return false;
        }
    }

    if (stat("/usr/bin/g++", &st) == 0) {
        if (st.st_mtime != gpp_bin_timestamp) {
            return false;
        }
    } else {
        if (gpp_bin_timestamp != 0) {
            return false;
        }
    }

    if (stat("/usr/bin/clang", &st) == 0) {
        if (st.st_mtime != clang_bin_timestamp) {
            return false;
        }
    } else {
        if (clang_bin_timestamp != 0) {
            return false;
        }
    }

    return true;
}

// Returns fd for icecc-create-env output
int start_create_env(const string &basedir, uid_t user_uid, gid_t user_gid,
                     const std::string &compiler, const list<string> &extrafiles)
{
    string nativedir = basedir + "/native/";

    if (compiler == "clang") {
        if (::access("/usr/bin/clang", X_OK) != 0) {
            return 0;
        }
    } else { // "gcc" (the default)
        // Both gcc and g++ are needed in the gcc case.
        if (::access("/usr/bin/gcc", X_OK) != 0 || ::access("/usr/bin/g++", X_OK) != 0) {
            return 0;
        }
    }

    if (mkdir(nativedir.c_str(), 0775) && errno != EEXIST) {
        return 0;
    }

    if (chown(nativedir.c_str(), user_uid, user_gid) ||
            chmod(nativedir.c_str(), 0775)) {
        log_perror("chown/chmod failed");
        if (-1 == rmdir(nativedir.c_str())){
            log_perror("rmdir failed");
        }
        return 0;
    }

    flush_debug();
    int pipes[2];
    if (pipe(pipes) == -1) {
        log_error() << "failed to create pipe: " << strerror(errno) << endl;
        _exit(147);
    }
    pid_t pid = fork();

    if (pid == -1) {
        log_perror("failed to fork");
        _exit(147);
    }

    if (pid) {
        if ((-1 == close(pipes[1])) && (errno != EBADF)){
            log_perror("close failed");
        }
        return pipes[0];
    }
    // else

#ifndef HAVE_LIBCAP_NG
    if (getuid() != user_uid || geteuid() != user_uid
	|| getgid() != user_gid || getegid() != user_gid) {

        if (setgroups(0, NULL) < 0) {
            log_perror("setgroups failed");
            _exit(143);
        }

        if (setgid(user_gid) < 0) {
            log_perror("setgid failed");
            _exit(143);
        }

        if (!geteuid() && setuid(user_uid) < 0) {
            log_perror("setuid failed");
            _exit(142);
        }
    }
#endif

    if (chdir(nativedir.c_str())) {
        log_perror("chdir") << "\t" << nativedir << endl;
        _exit(1);
    }

    if ((-1 == close(pipes[0])) && (errno != EBADF)){
        log_perror("close failed");
    }

    if (-1 == dup2(pipes[1], 5)){   // icecc-create-env will write the hash there
        log_perror("dup2 failed");
    }

    if ((-1 == close(pipes[1])) && (errno != EBADF)){
        log_perror("close failed");
    }

    if ((-1 == close(STDOUT_FILENO)) && (errno != EBADF)){ // hide output from icecc-create-env
        log_perror("close failed");
    }

    const char **argv;
    argv = new const char*[4 + extrafiles.size()];
    int pos = 0;
    argv[pos++] = BINDIR "/icecc";
    argv[pos++] = "--build-native";
    const int first_to_free = pos;
    argv[pos++] = strdup(compiler.c_str());

    for (list<string>::const_iterator it = extrafiles.begin(); it != extrafiles.end(); ++it) {
        argv[pos++] = strdup(it->c_str());
    }

    argv[pos++] = NULL;

    if (!exec_and_wait(argv)) {
        log_error() << BINDIR "/icecc --build-native failed" << endl;
        _exit(1);
    }
    for( int i = first_to_free; i < pos; ++i )
        free( (void*) argv[ i ] );
    delete[] argv;

    _exit(0);
}

size_t finish_create_env(int pipe, const string &basedir, string &native_environment)
{
// We don't care about waitpid() , icecc-create-env prints the name of the tarball as the very last
// action before exit, so if there's something in the pipe, just block on it until it closes.

    char buf[1024];
    buf[0] = '\0';

    while (read(pipe, buf, 1023) < 0 && errno == EINTR) {}

    if (char *nl = strchr(buf, '\n')) {
        *nl = '\0';
    }

    string nativedir = basedir + "/native/";
    native_environment = nativedir + buf;

    if ((-1 == close(pipe)) && (errno != EBADF)){
        log_perror("close failed");
    }
    trace() << "native_environment " << native_environment << endl;
    struct stat st;

    if (!native_environment.empty()
        && (stat(native_environment.c_str(), &st) == 0)) {
        return st.st_size;
    }

    if (-1 == rmdir(nativedir.c_str())){
        log_perror("rmdir failed");
    }
    return 0;
}


pid_t start_install_environment(const std::string &basename, const std::string &target,
                                const std::string &name, MsgChannel *c,
                                int &pipe_to_stdin, FileChunkMsg *&fmsg,
                                uid_t user_uid, gid_t user_gid, int extract_priority)
{
    if (!name.size()) {
        log_error() << "illegal name for environment " << name << endl;
        return 0;
    }

    for (string::size_type i = 0; i < name.size(); ++i) {
        if (isascii(name[i]) && !isspace(name[i]) && name[i] != '/' && isprint(name[i])) {
            continue;
        }

        log_error() << "illegal char '" << name[i] << "' - rejecting environment " << name << endl;
        return 0;
    }

    string dirname = basename + "/target=" + target;
    Msg *msg = c->get_msg(30);

    if (!msg || msg->type != M_FILE_CHUNK) {
        trace() << "Expected first file chunk\n";
        return 0;
    }

    fmsg = dynamic_cast<FileChunkMsg*>(msg);
    enum { BZip2, Gzip, None} compression = None;

    if (fmsg->len > 2) {
        if (fmsg->buffer[0] == 037 && fmsg->buffer[1] == 0213) {
            compression = Gzip;
        } else if (fmsg->buffer[0] == 'B' && fmsg->buffer[1] == 'Z') {
            compression = BZip2;
        }
    }

    if (mkdir(dirname.c_str(), 0770) && errno != EEXIST) {
        log_perror("mkdir target") << "\t" << dirname << endl;
        return 0;
    }

    if (chown(dirname.c_str(), user_uid, user_gid) || chmod(dirname.c_str(), 0770)) {
        log_perror("chown,chmod target") << "\t" << dirname << endl;
        return 0;
    }

    dirname = dirname + "/" + name;

    if (mkdir(dirname.c_str(), 0770)) {
        log_perror("mkdir name") << "\t" << dirname << endl;
        return 0;
    }

    if (chown(dirname.c_str(), user_uid, user_gid) || chmod(dirname.c_str(), 0770)) {
        log_perror("chown,chmod name") << "\t" << dirname << endl;
        return 0;
    }

    int fds[2];

    if (pipe(fds) == -1) {
        log_perror("pipe failed");
        return 0;
    }

    flush_debug();
    pid_t pid = fork();

    if (pid == -1) {
        log_perror("fork - trying to run tar");
        return 0;
    }
    if (pid) {
        trace() << "pid " << pid << endl;

        if ((-1 == close(fds[0])) && (errno != EBADF)){
            log_perror("close failed");
        }
        pipe_to_stdin = fds[1];

        return pid;
    }

    // else
#ifndef HAVE_LIBCAP_NG

    if (setgroups(0, NULL) < 0) {
        log_perror("setgroups fails");
        _exit(143);
    }

    if (setgid(user_gid) < 0) {
        log_perror("setgid fails");
        _exit(143);
    }

    if (!geteuid() && setuid(user_uid) < 0) {
        log_perror("setuid fails");
        _exit(142);
    }

#endif

    // reset SIGPIPE and SIGCHILD handler so that tar
    // isn't confused when gzip/bzip2 aborts
    signal(SIGCHLD, SIG_DFL);
    signal(SIGPIPE, SIG_DFL);

    if ((-1 == close(0)) && (errno != EBADF)){
        log_perror("close failed");
    }

    if ((-1 == close(fds[1])) && (errno != EBADF)){
        log_perror("close failed");
    }

    if (-1 == dup2(fds[0], 0)){
        log_perror("dup2 failed");
    }

    int niceval = nice(extract_priority);
    if (-1 == niceval){
        log_warning() << "failed to set nice value: " << strerror(errno) << endl;
    }

    char **argv;
    argv = new char*[6];
    argv[0] = strdup(TAR);
    argv[1] = strdup("-C");
    argv[2] = strdup(dirname.c_str());

    if (compression == BZip2) {
        argv[3] = strdup("-xjf");
    } else if (compression == Gzip) {
        argv[3] = strdup("-xzf");
    } else if (compression == None) {
        argv[3] = strdup("-xf");
    }

    argv[4] = strdup("-");
    argv[5] = 0;
    execv(argv[0], argv);
    log_perror("execv failed");
    _exit(100);
}


size_t finalize_install_environment(const std::string &basename, const std::string &target,
                                    pid_t pid, uid_t user_uid, gid_t user_gid)
{
    int status = 1;

    while (waitpid(pid, &status, 0) < 0 && errno == EINTR) {}

    if (shell_exit_status(status) != 0) {
        log_error() << "exit code: " << shell_exit_status(status) << endl;
        remove_environment(basename, target);
        return 0;
    }

    string dirname = basename + "/target=" + target;

    errno = 0;
    mkdir((dirname + "/tmp").c_str(), 01775);
    ignore_result(chown((dirname + "/tmp").c_str(), user_uid, user_gid));
    chmod((dirname + "/tmp").c_str(), 01775);
    if (errno == -1) {
        log_error() << "failed to setup " << dirname << "/tmp :"
                    << strerror(errno) << endl;
    }

    return sumup_dir(dirname);
}

size_t remove_environment(const string &basename, const string &env)
{
    string dirname = basename + "/target=" + env;

    size_t res = sumup_dir(dirname);

    flush_debug();
    pid_t pid = fork();

    if (pid == -1) {
        log_perror("failed to fork");
        return 0;
    }

    if (pid) {
        int status = 0;

        while (waitpid(pid, &status, 0) < 0 && errno == EINTR) {}

        if (WIFEXITED(status)) {
            return res;
        }

        // something went wrong. assume no disk space was free'd.
        return 0;
    }

    // else

    char **argv;
    argv = new char*[5];
    argv[0] = strdup("/bin/rm");
    argv[1] = strdup("-rf");
    argv[2] = strdup("--");
    argv[3] = strdup(dirname.c_str());
    argv[4] = NULL;

    execv(argv[0], argv);
    log_perror("execv failed");
    _exit(-1);
}

size_t remove_native_environment(const string &env)
{
    if (env.empty()) {
        return 0;
    }

    struct stat st;

    if (stat(env.c_str(), &st) == 0) {
        if (-1 == unlink(env.c_str())){
            log_perror("unlink failed") << "\t" << env << endl;
        }
        return st.st_size;
    }

    return 0;
}

static void
error_client(MsgChannel *client, string error)
{
    if (IS_PROTOCOL_23(client)) {
        client->send_msg(StatusTextMsg(error));
    }
}

void chdir_to_environment(MsgChannel *client, const string &dirname, uid_t user_uid, gid_t user_gid)
{
#ifdef HAVE_LIBCAP_NG

    if (chdir(dirname.c_str()) < 0) {
        error_client(client, string("chdir to ") + dirname + "failed");
        log_perror("chdir() failed") << "\t" << dirname << endl;
        _exit(145);
    }

    if (chroot(dirname.c_str()) < 0) {
        error_client(client, string("chroot ") + dirname + "failed");
        log_perror("chroot() failed") << "\t" << dirname << endl;
        _exit(144);
    }

    (void) user_uid;
    (void) user_gid;
#else

    if (getuid() == 0) {
        // without the chdir, the chroot will escape the
        // jail right away
        if (chdir(dirname.c_str()) < 0) {
            error_client(client, string("chdir to ") + dirname + "failed");
            log_perror("chdir() failed") << "\t" << dirname << endl;
            _exit(145);
        }

        if (chroot(dirname.c_str()) < 0) {
            error_client(client, string("chroot ") + dirname + "failed");
            log_perror("chroot() failed") << "\t" << dirname << endl;
            _exit(144);
        }

        if (setgroups(0, NULL) < 0) {
            error_client(client, string("setgroups failed"));
            log_perror("setgroups() failed");
            _exit(143);
        }

        if (setgid(user_gid) < 0) {
            error_client(client, string("setgid failed"));
            log_perror("setgid() failed");
            _exit(143);
        }

        if (setuid(user_uid) < 0) {
            error_client(client, string("setuid failed"));
            log_perror("setuid() failed");
            _exit(142);
        }
    } else {
        error_client(client, "cannot chroot to environment");
        _exit(146);
    }

#endif
}

// Verify that the environment works by simply running the bundled bin/true.
bool verify_env(MsgChannel *client, const string &basedir, const string &target, const string &env,
                uid_t user_uid, gid_t user_gid)
{
    if (target.empty() || env.empty()) {
        error_client(client, "verify_env: target or env empty");
        log_error() << "verify_env target or env empty\n\t" << target << "\n\t" << env << endl;
        return false;
    }

    string dirname = basedir + "/target=" + target + "/" + env;

    if (::access(string(dirname + "/bin/true").c_str(), X_OK)) {
        error_client(client, dirname + "/bin/true is not executable");
        log_error() << "I don't have environment " << env << "(" << target << ") to verify." << endl;
        return false;
    }

    flush_debug();
    pid_t pid = fork();
    assert(pid >= 0);

    if (pid > 0) {  // parent
        int status;

        while (waitpid(pid, &status, 0) < 0 && errno == EINTR) {}

        return shell_exit_status(status) == 0;
    } else if (pid < 0) {
        log_perror("fork failed");
        return false;
    }

    // child
    reset_debug();
    chdir_to_environment(client, dirname, user_uid, user_gid);
    execl("bin/true", "bin/true", (void*)NULL);
    log_perror("execl failed");
    _exit(-1);

}
